### ISSUE TITLE
J# language

### ISSUE THREAD

**Issue by KrzysztofSzewczyk** [github/linguist#4124]
Although J# is not-so-commonly used, it's still an language.
A few repositories:
 * https://github.com/matt77hias/jsl-smallpt
 * https://github.com/ankurdave/htw-3d
 * https://github.com/hikerlive/code4me/
 * https://github.com/duchien85/cellengine/blob/4bafd127f579d3cdc1f652500111ad8b88588b8e/gameedit/CellMusicEdit/AppGameBMS/Program.jsl
 * Many, many, many more: https://github.com/search?l=&q=extension%3Ajsl+Program&type=Code

I couldn't find more because it's really hard.

## Preliminary Steps

Please confirm you have...
- [X] reviewed [How Linguist Works](https://github.com/github/linguist#how-linguist-works),
- [X] reviewed the [Troubleshooting](https://github.com/github/linguist#troubleshooting) docs,
- [X] considered implementing an [override](https://github.com/github/linguist#overrides), (C# and Java alone can't fit there, it's just mix of these)
- [X] verified an issue has not already been logged for your issue ([linguist issues](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+repo%3Agithub/linguist)).

### URL of the affected repository:
As mentioned earlier, https://github.com/matt77hias/jsl-smallpt & https://github.com/ankurdave/htw-3d

### Last modified on:
2017-24-24, 2016-3-14

### Expected language:
J#

### Detected language:
None & "logos"

**Comment by pchaigno** [comment]
Are there other languages that you know of using the `.jsl` extension? Are there files in [this search query](https://github.com/search?utf8=%E2%9C%93&q=extension%3Ajsl+NOT+class+NOT+package&type=Code) that are not J#?

**Comment by KrzysztofSzewczyk** [comment]
Yes, some kind of shader language. It's different from J# and is C-Like, while J# has syntax like C#+Java.

**Comment by pchaigno** [comment]
Then we'll first need to estimate the number of `.jsl` files on Github for these two languages. Could you point me to a couple of these C-like files?

**Comment by KrzysztofSzewczyk** [comment]
Every file inside https://github.com/ojdkbuild/upstream_openjfx is written in C-like shader language.

1) https://github.com/ojdkbuild/upstream_openjfx/blob/2ccffe1ac8dfde4ccc819f5b4d189c0185823c85/modules/javafx.graphics/src/main/jsl-decora/Blend.jsl

2) https://github.com/ojdkbuild/upstream_openjfx/blob/2ccffe1ac8dfde4ccc819f5b4d189c0185823c85/modules/javafx.graphics/src/main/jsl-decora/Blend_ADD.jsl

3) https://github.com/ojdkbuild/upstream_openjfx/blob/2ccffe1ac8dfde4ccc819f5b4d189c0185823c85/modules/javafx.graphics/src/main/jsl-decora/Blend_BLUE.jsl

4) https://github.com/ojdkbuild/upstream_openjfx/blob/2ccffe1ac8dfde4ccc819f5b4d189c0185823c85/modules/javafx.graphics/src/main/jsl-decora/Blend_COLOR_BURN.jsl

**Comment by KrzysztofSzewczyk** [comment]
From my observatons, this files contain not-OO code, so I think it's possible to just search for C#+Java like only tokens.

If it's needed, I can hand some samples and release them to public domain

**Comment by pchaigno** [comment]
@KrzysztofSzewczyk Using [the `extension:jsl float4 OR param` search query](https://github.com/search?utf8=%E2%9C%93&q=extension%3Ajsl+float4+OR+param&type=) and [Harvester](https://github.com/Alhadis/Harvester), I count only 48 repositories by 42 users. That's not enough to warrant support in Linguist, so we won't need to add this language.

For J#, Harvester finds 398 repositories by 366 in the 2156 files I downloaded from [the `extension:jsl class OR private OR public OR package` search query](https://github.com/search?utf8=%E2%9C%93&q=extension%3Ajsl+class+OR+private+OR+public+OR+package&type=Code). That's enough to warrant support in Linguist. We'll gladly accept [a pull request to add that language](https://github.com/github/linguist/blob/master/CONTRIBUTING.md#adding-a-language) :smiley:

**Comment by KrzysztofSzewczyk** [comment]
In which format should I provide grammar for language?

**Comment by pchaigno** [comment]
Linguist (and GitHub.com) support Sublime Text, Atom, or TextMate grammars.

**Comment by KrzysztofSzewczyk** [comment]
Syntax is actually same as C# (some exceptions are present, tho), so where should I look for eg. C#/Java grammars?

**Comment by pchaigno** [comment]
Those we already use should be fine: https://github.com/github/linguist/tree/master/vendor.

**Comment by KrzysztofSzewczyk** [comment]
I ran to some kind of strange problem:
```
$ script/add-grammar https://github.com/KrzysztofSzewczyk/JSharpSyntax.git
Command failed. Aborting.
Registering new submodule: vendor/grammars/JSharpSyntax
$ git submodule add -f https://github.com/KrzysztofSzewczyk/JSharpSyntax vendor/grammars/JSharpSyntax
  > Cloning into 'C:/Users/PC/Desktop/linguist/vendor/grammars/JSharpSyntax'...
  > fatal: clone of 'https://github.com/KrzysztofSzewczyk/JSharpSyntax' into submodule path 'C:/Users/PC/Desktop/linguist/vendor/grammars/JSharpSyntax' failed
```

Propably I'm missing something. When I'll remove ".git" it doesn't work too.

**Comment by KrzysztofSzewczyk** [comment]
@pchaigno I can use C# grammar, but the syntax highlighting won't be 100% accurate.

**Comment by pchaigno** [comment]
I'm asking because it looked like you were [simply reusing the Java grammar](https://github.com/KrzysztofSzewczyk/JSharpSyntax/commits/master). To use any grammar already used by another language, you can just fill out `tm_scope: scope_of_that_grammar` (e.g., `tm_scope: source.java`) in your `languages.yml` entry.

**Comment by KrzysztofSzewczyk** [comment]
OK, I've done it. `script/set-language-ids` fails somehow (I couldnt find any depedency-installing script).
The error is:  "Could not find gem 'rake-compiler (~> 0.9) x64-mingw32' in any of the gem sources listed in your Gemfile. (Bundler::GemNotFound)"
I've already tried `gem install rake-compiler`, without luck. I'm total newbie in Ruby, so please forgive me if it's stupid question.

**Comment by pchaigno** [comment]
@KrzysztofSzewczyk Did you make progress on this? I'm sorry I can't be of much help for the above problem, I've never installed Linguist on Windows.

**Comment by KrzysztofSzewczyk** [comment]
I got a stuck there, couldn't find help from friends programming in Ruby, then I forgot about it.
Now I have Linux machine so I'm going to test it out

**Comment by KrzysztofSzewczyk** [comment]
![Screenshot](https://user-images.githubusercontent.com/27734421/44383689-d6203300-a519-11e8-99da-4132074fb2fc.png)
I have problem with update-ids once again, it assigns some meaningless number to language...

**Comment by pchaigno** [comment]
That's how it's supposed to work. As long as it only adds an id to J#, it should be alright.

**Comment by KrzysztofSzewczyk** [comment]
So it's OK then?

**Comment by pchaigno** [comment]
Not all of them. It's okay.

**Comment by KrzysztofSzewczyk** [comment]
[PR](https://github.com/github/linguist/pull/4242) was made, so I'm closing issue as it's not needed currently.
